cpp/include/cudf/strings/detail/json_parser.cuh

```diff
@@ -0,0 +1,970 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include <thrust/optional.h>
+#include <rmm/device_vector.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * Json token enum
+ */
+enum class json_token {
+  // start token
+  INIT = 0,
+
+  // successfully parsed the whole Json string
+  SUCCESS,
+
+  // get error when parsing Json string
+  ERROR,
+
+  // '{'
+  START_OBJECT,
+
+  // '}'
+  END_OBJECT,
+
+  // '['
+  START_ARRAY,
+
+  // ']'
+  END_ARRAY,
+
+  // e.g.: key1 in {"key1" : "value1"}
+  FIELD_NAME,
+
+  // e.g.: value1 in {"key1" : "value1"}
+  VALUE_STRING,
+
+  // e.g.: 123 in {"key1" : 123}
+  VALUE_NUMBER_INT,
+
+  // e.g.: 1.25 in {"key1" : 1.25}
+  VALUE_NUMBER_FLOAT,
+
+  // e.g.: true in {"key1" : true}
+  VALUE_TRUE,
+
+  // e.g.: false in {"key1" : false}
+  VALUE_FALSE,
+
+  // e.g.: null in {"key1" : null}
+  VALUE_NULL
+
+};
+
+/**
+ * options for json_parser
+ *
+ */
+class json_parser_options {
+ public:
+  CUDF_HOST_DEVICE inline bool get_allow_single_quotes() const { return allow_single_quotes; }
+
+  CUDF_HOST_DEVICE inline bool get_allow_unescaped_control_chars() const
+  {
+    return allow_unescaped_control_chars;
+  }
+
+  CUDF_HOST_DEVICE inline bool get_max_nesting_depth() const { return max_nesting_depth; }
+
+  void set_allow_single_quotes(bool _allow_single_quotes)
+  {
+    allow_single_quotes = _allow_single_quotes;
+  }
+
+  void set_allow_unescaped_control_chars(bool _allow_unescaped_control_chars)
+  {
+    allow_unescaped_control_chars = _allow_unescaped_control_chars;
+  }
+
+ private:
+  // if true, allow both ' and " for quoting strings. e.g.: json string {'k1' : "v1"} is valid
+  // if true, Json is not valid formal format
+  bool allow_single_quotes = false;
+
+  // Feature that determines whether parser will allow JSON Strings to contain unescaped control
+  // characters (ASCII characters with value less than 32, including tab and line feed characters)
+  // or not.
+  // e.g., how to represent carriage return and newline character characters:
+  //   if false, "\\n\\r" is allowed. There are 4 chars(\ n \ r), after escaped, 2 chars remains.
+  //     Note: print "\\n\\r" => \n\r
+  //   if true, allow "\n\r" string
+  //     Note: in this string only contains 2 charactors:
+  //       carriage return character (ASCII code 13)
+  //       newline character (ASCII code 10)
+  bool allow_unescaped_control_chars = false;
+
+  // it's from Jackson json
+  int max_nesting_depth = 1000;
```

Comment:
Updated logic:
Use a template parameter to define `max_nesting_depth` in compile time instead of runtime time.
```
json_parser<max_json_nesting_depth> j_parser(j_parser_options, str.data(), str.size_bytes());
```